File: src/cli/commands/down.ts
/** `jackops down` command handler. */

import * as tmux from "../../tmux.ts";
import * as worktree from "../../worktree.ts";
import { sessionName } from "../../config.ts";
import { INIT_SESSION } from "../../init.ts";
import { loadCliConfig, prompt } from "../context.ts";

export async function down(): Promise<void> {
  const base = Deno.cwd();
  let project: string | undefined;

  // Try loading config for project name, but don't require it
  try {
    const { config } = await loadCliConfig();
    project = config.project;
  } catch {
    // Config missing -- fall back to cleaning up all jackops worktrees
  }

  if (project) {
    const session = sessionName(project);
    if (await tmux.hasSession(session)) {
      await tmux.killSession(session);
      console.log(`Killed tmux session '${session}'.`);
    } else {
      console.log(`No active session '${session}'.`);
    }
  } else {
    console.log("No config found. Skipping tmux session cleanup.");
  }

  // Also kill init session if it exists
  if (await tmux.hasSession(INIT_SESSION)) {
    await tmux.killSession(INIT_SESSION);
    console.log(`Killed init session '${INIT_SESSION}'.`);
  }

  const entries = await worktree.list(base);
  const jackopsWorktrees = entries.filter((e) =>
    worktree.isJackopsWorktree(e, project)
  );

  if (jackopsWorktrees.length === 0) {
    console.log("No worktrees to clean up.");
    return;
  }

  const dirtyCounts = await Promise.all(
    jackopsWorktrees.map(async (e) => ({
      entry: e,
      dirty: await worktree.dirtyCount(e.path),
    })),
  );
  const anyDirty = dirtyCounts.some((d) => d.dirty > 0);

  if (anyDirty) {
    console.log(`\nWorktrees to remove:`);
    for (const { entry, dirty } of dirtyCounts) {
      console.log(`  ${entry.path} (${worktree.formatDirtyLabel(dirty)})`);
    }

    const answer = await prompt(
      "\nWorktrees contain uncommitted changes. Remove anyway? [y/N] ",
    );
    if (answer !== "y") {
      console.log("Worktrees kept.");
      return;
    }
  }

  const removed = await worktree.cleanup(
    base,
    project ?? "",
    jackopsWorktrees,
  );
  console.log(
    `Removed ${removed.length} worktree${removed.length !== 1 ? "s" : ""}.`,
  );
}

File: src/cmd/down.go
package cmd

import (
	"fmt"
	"os"

	"github.com/AdriaanVE/jack-in/src/config"
	"github.com/AdriaanVE/jack-in/src/tmux"
	"github.com/AdriaanVE/jack-in/src/worktree"
	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(downCmd)
}

var downCmd = &cobra.Command{
	Use:   "down",
	Short: "Kill session and clean up worktrees",
	RunE:  runDown,
}

func runDown(cmd *cobra.Command, args []string) error {
	base := cwd()

	// Try loading config for project name, but don't require it
	var project string
	cfg, err := loadConfig(base)
	if err == nil {
		project = cfg.Project
	}

	// Kill project session
	if project != "" {
		session := config.SessionName(project)
		if tmux.HasSession(session) {
			if err := tmux.KillSession(session); err != nil {
				return fmt.Errorf("killing session: %w", err)
			}
			fmt.Printf("Killed tmux session '%s'.\n", session)
		} else {
			fmt.Printf("No active session '%s'.\n", session)
		}
	} else {
		fmt.Println("No config found. Skipping tmux session cleanup.")
	}

	// Kill init session if it exists
	if tmux.HasSession(initSessionName) {
		if err := tmux.KillSession(initSessionName); err != nil {
			return fmt.Errorf("killing init session: %w", err)
		}
		fmt.Printf("Killed init session '%s'.\n", initSessionName)
	}

	// Find jackin worktrees
	allWorktrees, err := worktree.List(base)
	if err != nil {
		return fmt.Errorf("listing worktrees: %w", err)
	}
	jackinWTs := filterJackInWorktrees(allWorktrees, project)

	if len(jackinWTs) == 0 {
		fmt.Println("No worktrees to clean up.")
		return nil
	}

	// Check for dirty worktrees and prompt before removing
	fmt.Println("\nWorktrees to remove:")
	anyDirty := printWorktreeStatuses(jackinWTs)
	if anyDirty {
		answer := promptUser("\nWorktrees contain uncommitted changes. Remove anyway? [y/N] ")
		if answer != "y" {
			fmt.Println("Worktrees kept.")
			return nil
		}
	}

	removed, errs := worktree.RemoveAll(base, jackinWTs)
	for _, e := range errs {
		fmt.Fprintf(os.Stderr, "  %v\n", e)
	}
	fmt.Printf("Removed %d worktree%s.\n", len(removed), plural(len(removed)))
	return nil
}

func plural(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}
